polygeist/lib/Dialect/Polygeist/Analysis/MemoryAccessAnalysis.cpp

```diff
@@ -641,6 +676,277 @@ MemoryAccess<OpTy>::classifyMemoryAccess(DataFlowSolver &solver) const {
+  //  - base operand equal to the result of a sycl accessor subscript, and
+  //  - a single index with zero value.
+  MemRefAccess access(memoryOp);
+  if (!isa<sycl::SYCLAccessorSubscriptOp>(access.memref.getDefiningOp()) ||
+      !hasZeroIndex(access))
+    return;
+
+  LLVM_DEBUG(llvm::errs() << "Candidate op:" << memoryOp << "\n");
+
+  auto accessorSubscriptOp =
+      cast<sycl::SYCLAccessorSubscriptOp>(access.memref.getDefiningOp());
+  Value accSubIndex = accessorSubscriptOp.getIndex();
+
+  // Try to determine the underlying value(s) of the accessor subscript index
+  // operand. The number of underlying values should be equal to the
+  // dimensionality of the sycl id used as an index in the accessor subscript
+  // operation. Example:
+  //   sycl.constructor @id(%id, %i, %j) : (memref<?x!sycl_id_2>, i64, i64)
+  //   %sub = sycl.accessor.subscript %acc[%id] ...
+  // Here the underlying values for '%id' are {%i, %j}. The number of underlying
+  // values matches the dimensionality of the sycl id.
+  SmallVector<Value> underlyingVals = getUnderlyingValues(
+      accessorSubscriptOp.getOffsetOperandIndex(), accessorSubscriptOp, solver);
+  if (underlyingVals.empty()) {
+    LLVM_DEBUG(llvm::dbgs() << "Failed to resolve underlying value(s) for: "
+                            << accessorSubscriptOp.getIndex() << "\n");
+    return;
+  }
+
+  assert(
+      sycl::getDimensions(accSubIndex.getType()) == underlyingVals.size() &&
+      "Number of underlying values should be equal to dimensionality of the id "
+      "used to index the accessor");
+  LLVM_DEBUG({
+    for (Value val : underlyingVals)
+      llvm::dbgs() << "Underlying val: " << val << "\n";
+  });
+
+  // Construct the memory access matrix. The number of rows is equal to the
+  // dimensionality of the sycl.id used by the accessor subscript operation. The
+  // number of columns is equal to the number of loops surrounding the memory
+  // access plus (TODO) the dimensionality of the sycl item received by the
+  // kernel.
+  SetVector<Operation *> enclosingLoops =
+      getParentsOfType<AffineForOp>(memoryOp->getBlock());
+  MemoryAccessMatrix accessMatrix(sycl::getDimensions(accSubIndex.getType()),
+                                  enclosingLoops.size());
+
+  // Remove conversion like operations in the incoming expression if any.
+  auto removeConversions = [](Value expr) {
+    if (matchPattern(expr, m_Op<arith::ExtUIOp>()) ||
+        matchPattern(expr, m_Op<arith::ExtSIOp>()) ||
+        matchPattern(expr, m_Op<arith::TruncIOp>()) ||
+        matchPattern(expr, m_Op<arith::IndexCastOp>()) ||
+        matchPattern(expr, m_Op<arith::IndexCastUIOp>()))
+      return expr.getDefiningOp()->getOperand(0);
+    return expr;
+  };
+
+  // Determine whether 'expr' is a multiplication involving 'val'.
+  auto getMultiplier = [&](Value expr, Value val) -> std::optional<Value> {
+    expr = removeConversions(expr);
```

Comment:
do we need to removeConversions for val as well?

Reply:
not sure what to see, but resolving the coversation.

Reply:
The link is gone now, but I was referencing to this comment:
```
Not sure, this (and getMultiplier) is a part of the code that is WIP as I mentioned in the PR abstract but I thought to put out a PR for review now and build on it in part2, 3, ...
```